nixos/modules/tasks/filesystems/overlayfs.nix

```diff
@@ -0,0 +1,151 @@
+{ config, lib, pkgs, utils, ... }:
+
+let
+  toplevelConfig = config;
+
+  # Returns a service that creates the required directories before the mount is
+  # created.
+  preMountService = _name: fs:
+    let
+      rwName = "rw-${utils.escapeSystemdPath fs.mountPoint}";
+
+      # The scripted initrd contains some magic to add the prefix to the
+      # paths just in time, so we don't add it here.
+      prefix =
+        if config.boot.initrd.systemd.enable && fs.neededForBoot then
+          "/sysroot"
+        else
+          "";
+
+      upperdir = prefix + fs.overlay.upperdir;
+      workdir = prefix + fs.overlay.workdir;
+    in
+    lib.mkIf (fs.overlay.upperdir != null)
+      {
+        "${rwName}" = {
+          # Wait for sysroot to be mounted before creating upperdir and
+          # workdir.
+          after = [ "sysroot.mount" ];
+          unitConfig = {
+            DefaultDependencies = false;
+          };
+          serviceConfig = {
+            Type = "oneshot";
+            ExecStart = "${pkgs.coreutils}/bin/mkdir -p -m 0755 ${toString upperdir} ${toString workdir}";
```

Comment:
Don't think `toString` is necessary here, is it?

Reply:
Should the mode `0755` be configurable with like `fileSystems.<name>.overlay.mode` or something?

Reply:
Maybe? I think `overlay.mode` would suggest that you somehow set the mode of the overlay instead (as opposed to the directories backing the writable layer). Also not sure if changing the mode is allowed or sensible. I'd book this a future improvement if we find out that we need/want it.

Reply:
I'd say guarding against accidental import-into-store explicitly with toString makes sense? Or am I missing a guarantee that it won't be a store path elsewhere?

Reply:
> I'd say guarding against accidental import-into-store explicitly with toString makes sense

This conversation demonstrates pretty well why I think `types.path` is not a good choice unless you actually want to copy a path to the store.

Reply:
Switched upperdir and workdir to `types.str`